designs/opencl_kernel_generator.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: opencl_kernel_generator
+- Start Date: 2019-10-21
+- RFC PR: (leave this empty)
+- Stan Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Define operators and functions on matrix_cl that will use 
+[expresson templates](https://en.wikipedia.org/wiki/Expression_templates) 
+to construct expressions. When assigned to a matrix_cl an expression will generate an 
+OpenCL kernel an execute it to calculate the result of the expression.
+
+# Motivation
+[motivation]: #motivation
+
+Using expression templates is much simpler than writing OpenCL kernels by hand and it does 
+not require knowledge of OpenCL and parallel programming.
+
+Using one kernel for multiple operations is faster than using one kernel per operation. Each 
+kernel must load its arguments from global global GPU memory and store results back. Memory 
+transfers are relatively slow compared to calculations. In case of simple kernels majority 
+of kernel execution time is spent for memory transfers.  
```

Comment:
> I though the global memory in OpenCL referred to the Device memory (ala DRAM for the GPU) so the real transfer cost is in first moving the data over to DRAM.

That is correct. However, kernel generator does not modify anything about how much data is moved between CPU and GPU RAM.

>Because at that point it's the function call from Host -> Device and the creation of temporaries for intermediate operations that is the additional cost this cuts out.

Allocating memory for temporaries is not the slow part. GPU reading and writing them to global memory is. While GPU global memory achieves larger transfere speeds than RAM CPU is using, it is still not as much faster than GPU is compared to CPU. The result is GPU can do far more calculations in the same time it does a memory transfer than CPU. Effectively that means in same algorithm larger precentage of time is spent doing memory transfers.